OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,448 @@
+            }
+
+            if (CurrentState == ExecutionState.Stopped)
+            {
+                _logger.Log(LogLevel.WARN, "ODP is not enabled.");
+                return;
+            }
+
+            if (InvalidDataFound(odpEvent.Data))
+            {
+                _logger.Log(LogLevel.ERROR, "ODP data is not valid.");
+                return;
+            }
+
+            odpEvent.Data = AugmentCommonData(odpEvent.Data);
+            Enqueue(odpEvent);
+        }
+
+        /// <summary>
+        /// Add a new ODP event to the queue
+        /// </summary>
+        /// <param name="odpEvent"></param>
+        private void Enqueue(OdpEvent odpEvent)
+        {
+            if (_queue.Count >= _queueSize)
+            {
+                _logger.Log(LogLevel.WARN, $"ODP event send failed (queueSize = {_queue.Count}).");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
+
+            ProcessQueue();
+        }
+
+        /// <summary>
+        /// Process the queue
+        /// </summary>
+        private void ProcessQueue()
+        {
+            if (CurrentState != ExecutionState.Running)
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Processing Queue.");
+
+            lock (lockObject)
+            {
+                CurrentState = ExecutionState.Processing;
+                while (QueueHasBatches())
+                {
+                    DequeueSendSingleBatch();
+                }
+
+                CurrentState = ExecutionState.Running;
+            }
+        }
+
+        private void RegularlyFlushQueue()
+        {
+            while (!_flushIntervalCancellation.IsCancellationRequested)
+            {
+                Task.Delay(_flushInterval).ContinueWith(_ => FlushQueue()).Wait();
+            }
+        }
+
+        private void FlushQueue()
+        {
+            if (CurrentState != ExecutionState.Running)
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Flushing Queue.");
+
+            lock (lockObject)
+            {
+                CurrentState = ExecutionState.Processing;
+                while (QueueContainsItems())
+                {
+                    DequeueSendSingleBatch();
+                }
+
+                CurrentState = ExecutionState.Running;
+            }
+        }
+
+        /// <summary>
+        /// Dequeue a single batch and send it
+        /// </summary>
+        private void DequeueSendSingleBatch()
+        {
+            var batch = new List<OdpEvent>(_batchSize);
+
+            // dequeue a batch 
+            for (int i = 0; i < _batchSize && _queue.Count > 0; i += 1)
+            {
+                if (_queue.TryDequeue(out OdpEvent dequeuedOdpEvent))
+                {
+                    batch.Add(dequeuedOdpEvent);
+                }
+            }
+
+            if (batch.Count > 0)
+            {
+                Task.Run(() =>
```

Comment:
Do we want to create a thread again here? This will create multiple concurrent `_odpEventApiManager.SendEvents`.